libraries/shared/watcher/event_watcher.go

```diff
@@ -17,137 +17,114 @@
-		if err != nil {
-			logrus.WithFields(logrus.Fields{
-				"headerId":    header.Id,
-				"headerHash":  header.Hash,
-				"blockNumber": header.BlockNumber,
-			}).Errorf("Couldn't fetch logs for header: %v", err)
-			return err
-		}
-
-		transactionsSyncErr := watcher.Syncer.SyncTransactions(header.Id, logs)
-		if transactionsSyncErr != nil {
-			logrus.Errorf("error syncing transactions: %s", transactionsSyncErr.Error())
-			return transactionsSyncErr
+// Extracts and delegates watched log events.
+func (watcher *EventWatcher) Execute(recheckHeaders constants.TransformerExecution, errsChan chan error) {
+	extractErrsChan := make(chan error)
+	delegateErrsChan := make(chan error)
+
+	go watcher.extractLogs(recheckHeaders, extractErrsChan)
+	go watcher.delegateLogs(delegateErrsChan)
+
+	for {
+		select {
+		case extractErr := <-extractErrsChan:
+			logrus.Errorf("error extracting logs in event watcher: %s", extractErr.Error())
+			errsChan <- extractErr
+		case delegateErr := <-delegateErrsChan:
+			logrus.Errorf("error delegating logs in event watcher: %s", delegateErr.Error())
+			errsChan <- delegateErr
 		}
+	}
+}
 
-		transformErr := watcher.transformLogs(logs, header)
-		if transformErr != nil {
-			logrus.Error("Could not transform logs: ", transformErr)
-			return transformErr
+func (watcher *EventWatcher) extractLogs(recheckHeaders constants.TransformerExecution, errs chan error) {
+	extractLogsErr := make(chan error)
+	missingHeadersFound := make(chan bool)
+	go watcher.LogExtractor.ExtractLogs(recheckHeaders, extractLogsErr, missingHeadersFound)
+
+	for {
+		select {
+		case err := <-extractLogsErr:
+			errs <- err
+		case missingHeaders := <-missingHeadersFound:
+			if missingHeaders {
+				go watcher.extractLogs(recheckHeaders, errs)
+			} else {
+				time.Sleep(NoNewDataPause)
+				go watcher.extractLogs(recheckHeaders, errs)
+			}
 		}
 	}
-	return err
 }
 
-func (watcher *EventWatcher) transformLogs(logs []types.Log, header core.Header) error {
-	chunkedLogs := watcher.Chunker.ChunkLogs(logs)
-
-	// Can't quit early and mark as checked if there are no logs. If we are running continuousLogSync,
-	// not all logs we're interested in might have been fetched.
-	for _, t := range watcher.Transformers {
-		transformerName := t.GetConfig().TransformerName
-		logChunk := chunkedLogs[transformerName]
-		err := t.Execute(logChunk, header)
-		if err != nil {
-			logrus.Errorf("%v transformer failed to execute in watcher: %v", transformerName, err)
-			return err
+func (watcher *EventWatcher) delegateLogs(errs chan error) {
+	delegateLogsErr := make(chan error)
+	logsFound := make(chan bool)
+	go watcher.LogDelegator.DelegateLogs(delegateLogsErr, logsFound)
+
+	for {
+		select {
+		case err := <-delegateLogsErr:
+			errs <- err
+		case logs := <-logsFound:
+			if logs {
+				go watcher.delegateLogs(errs)
+			} else {
+				time.Sleep(NoNewDataPause)
+				go watcher.delegateLogs(errs)
```

Comment:
Maybe I'm just tired, but doesn't this spawn an ever increasing family of goroutines since every call to `delegateLogs` runs a while loop?

Reply:
Probably need to think about this more, but my intention was that `logsFound` would only get hit when `delegateLogs` had exited - so you should only have one `delegateLogs` routine running at a time

Reply:
`extractLogs` and `delegateLogs` look like they never exit their for loops so I think you do end up spawning an indefinite number of goroutines . `ExtractLogs` and `DelegateLogs` exit after sending to their `missingHeadersFound` and `logsFound` channels, respectively, and that triggers their parent `extractLogs` or `delegateLogs` contexts to recursively spawn a new `extractLogs` or `delegateLogs` goroutine without exiting the current context (now left in for loop waiting on channels that have no potential senders).  I think just need to add a `return` after the the recursive calls. Maybe missing something though.

Reply:
Also not sure it makes sense to have this asynchronousity within the `extractLogs` and `delegateLogs` methods. The for loops in both methods are waiting for the `ExtractLogs` or `DelegateLogs` to return either an error or the `missingHeaders`/`logs` bool before they continue, and those methods don't cycle they only return one value before they exit, so it seems like this could be subbed with a synchronous method call and return.

Reply:
Got it - I think I follow why this is spawning tons of goroutines and why adding a `return` after the recursive call would help. Not sure I follow how this would be restructured if `extractLogs` and `delegateLogs` didn't fire off their own goroutines? Would the caller (`Execute`) be responsible for triggering new calls when they exit on `missingHeadersFound`/`logsFound`? Just trying to make sure the process doesn't end after all existing headers have been checked (when new headers are still coming in)...

Reply:
@m0ar @i-norden thanks for pointing this out! Pushed up some changes to reduce the amount of goroutines we're firing here - curious to collect your thoughts

Reply:
Changes look great! I can't spot any issues.